crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,289 @@
+{
+    size_t i;
+
+    /*
+     * If there is an underlying legacy key and it has changed, invalidate
+     * the cache of provider keys.
+     */
+    if (pk->pkey.ptr != NULL) {
+        /*
+         * If there is no dirty checker, this key can't be used
+         * with providers
+         */
+        if (pk->ameth->is_dirty == NULL)
+            return NULL;
+
+        if (pk->ameth->is_dirty(pk))
+            evp_keymgmt_clear_pkey_cache(pk);
+    }
+
+    /*
+     * See if we have exported to this provider already.
+     * If we have, return immediately.
+     */
+    for (i = 0;
+         i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+         i++) {
+        if (keymgmt == pk->pkeys[i].keymgmt)
+            return pk->pkeys[i].provkey;
+    }
+
+    /*
+     * If there are no cached provider keys, try to export a legacy key
+     * to the provider.
+     */
+    if (i == 0) {
+        /* If there is no legacy key, give up */
+        if (pk->pkey.ptr == NULL)
+            return NULL;
+
+        /* Also, if there is no export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it and return the result */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
+        pk->pkeys[0].provkey = pk->ameth->export_to(pk, keymgmt);
+        return pk->pkeys[0].provkey;
+    }
+
+    /*
+     * There are cached provider keys, so import from the first one
+     * and export the imported data to the new provider.
+     * If it's impossible to import from the cached provider, give up
+     * (transfering keys between providers can't be supported in that
+     * case)
+     */
+    if (keymgmt->importkey_priv != NULL
+        && pk->pkeys[0].keymgmt->exportkey_priv != NULL)
+    {
+        const OSSL_PARAM *paramdefs = NULL;
+        OSSL_PARAM *params = NULL;
+        void *data = NULL;
+        void *provctx = ossl_provider_ctx(EVP_KEYMGMT_provider(keymgmt));
+
+        paramdefs = pk->pkeys[0].keymgmt->exportkey_priv_types();
+        /*
+         * All params have 'data' set to NULL.  In that case, the exportkey
+         * call should just fill in 'return_size' in all applicable params.
+         */
+        params = paramdefs_to_params(paramdefs);
+        /* Get 'return_size' filled */
+        pk->pkeys[0].keymgmt->exportkey_priv(pk->pkeys[0].provkey, params);
+
+        /* Allocate space and assign 'data' to point into the data block */
+        data = allocate_params_space(params);
+
+        /* Call the exportkey function a second time, to get the data filled */
+        pk->pkeys[0].keymgmt->exportkey_priv(pk->pkeys[0].provkey, params);
+
+        /*
+         * We should have all the data at this point, so import into the
+         * new provider and hope to get a key back.
+         */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
```

Comment:
Shouldn't this be using `i` here rather than 0?

Reply:
Ah, yes, that's right.

Reply:
Fixed.